### ISSUE TITLE
Emoji support

### ISSUE THREAD

**Issue by HakanKaraoglu** [SolrNet/SolrNet#571]
I am trying to index the following twitter data in two different ways. The first option is a simple index option with the SolrNet library. When I do this the emojis don't show up properly in solr.

Tweet: "aşı oluyorum 😀😁😍😆😟😱😼😻🙌😀🤲😹🧠🤤 test yenisi"

![image](https://user-images.githubusercontent.com/12962730/134675829-d52ea9f8-f793-4617-a910-d12c4c55d684.png)

```
           var solrDocument = new SolrDocument();

            solrDocument.Id = "440550419";
            solrDocument.Title = "aşı oluyorum 😀😁😍😆😟😱😼😻🙌😀🤲😹🧠🤤 test yenisi";
            solrDocument.Body = "aşı oluyorum 😀😁😍😆😟😱😼😻🙌😀🤲😹🧠🤤 test yenisi";
            solrDocument.WorkspaceId = 114;

            var url = $"http://localhost:8983/solr/2021";

            var connection = new SolrConnection(url);

            Startup.Init<SolrDocument>(connection);
            var solr = ServiceLocator.Current.GetInstance<ISolrOperations<SolrDocument>>();

            solr.Add(solrDocument);
```

however, I noticed that when I send the json with an httpclient, which is the other option, it looks fine in solr.

```
 var solrDocument = new SolrDocument();

            solrDocument.Id = "440550419";
            solrDocument.Title = "aşı oluyorum 😀😁😍😆😟😱😼😻🙌😀🤲😹🧠🤤 test yenisi";
            solrDocument.Body = "aşı oluyorum 😀😁😍😆😟😱😼😻🙌😀🤲😹🧠🤤 test yenisi";
            solrDocument.WorkspaceId = 114;

var list = new List<SolrDocument>();
            list.Add(solrDocument);

            var json = JsonConvert.SerializeObject(list);

            var _client = new HttpClient();
            var response = await _client.PostAsync("http://localhost:8983/solr/2021/update?commitWithin=1000", new StringContent(json, Encoding.UTF8, MediaTypeNames.Application.Json));
```
![image](https://user-images.githubusercontent.com/12962730/135232660-c1da24f5-2607-4736-adc2-b998e988abb6.png)

How can I fix this problem? So somehow the solrnet library can corrupt these emojis?

Solr version: 7.1.0
SolrNet version: 1.0.19

**Comment by HakanKaraoglu** [comment]
@mausch Do you have any guidance on this?

**Comment by mausch** [comment]
I just [added a test that passed](https://github.com/SolrNet/SolrNet/commit/f32f6735b5867a8561929e4c7b684967a4fe6c9f).
Before that I've had [similar tests in codebase for over 10 years](https://github.com/SolrNet/SolrNet/commit/06d88e10c0b4d7970fe59902518d0e5a62b63d15).

**Comment by HakanKaraoglu** [comment]
> I just [added a test that passed](https://github.com/SolrNet/SolrNet/commit/f32f6735b5867a8561929e4c7b684967a4fe6c9f). Before that I've had [similar tests in the codebase for over 10 years](https://github.com/SolrNet/SolrNet/commit/06d88e10c0b4d7970fe59902518d0e5a62b63d15).

Yes, the test passes, but it does not index properly on the solr side, this is what I want to talk about. Did you test like this? I tested it and the result is as follows. Since we collect content from social media, emojis cause problems for us, so I wanted to open this topic.

  ```
const string name = "aşı 😀😁😍😆😟😱😼😻🙌😀🤲😹🧠🤤";
            await solr.AddAsync(new SolrDocument
            {
                Id = "440550422",
                Title = name
            });

            await solr.CommitAsync();
```

![image](https://user-images.githubusercontent.com/12962730/135358134-383e5748-f4b9-4721-aaee-1cb75d7d3bbe.png)

**Comment by HakanKaraoglu** [comment]
@mausch 
Can you test what I wrote? You will see that no emojis are displayed in Solr. This code may pass the test, but it's okay if it doesn't show emojis?

**Comment by hoerup** [comment]
HakanKaraoglu just to make sure: 
You do indeed have successfully added it directly to solr and got it back and thus verified that it is handled correctly in solr itself ?

**Comment by HakanKaraoglu** [comment]
The code I wrote for the test is as above. If you see a mistake, I can correct it and try again.

**Comment by hoerup** [comment]
My point is: In a setup with a client-library(SolrNet) and a server(solr) it can seem unfair to blame the client without verifying that the server is doing what it's supposed to do

so have you tried to inject a sample doc with emojis into solr either via solr UI or the cli tool <solrhome>/bin/post ?

**Comment by HakanKaraoglu** [comment]
yes, emojis appear when I update to the same document from the Solr UI. Even emojis appear when I throw them with http client.

**Comment by hoerup** [comment]
@mausch HakanKaraoglu might have a point
I tried to update the AddAndQueryUnicode test and add his list of emojis - and it fails : https://github.com/hoerup/SolrNet/actions/runs/1304932806

**Comment by mausch** [comment]
Yep, I have also confirmed the issue a couple of days ago... very strange!

**Comment by HakanKaraoglu** [comment]
@mausch Maybe this problem can be fixed with a new version. 
@hoerup Thank you for supporting my question. :)

**Comment by HakanKaraoglu** [comment]
@mausch any update ?

**Comment by omidontop** [comment]
Any updates on this issue?

**Comment by mausch** [comment]
Currently accepting PRs with a test+fix for this.

**Comment by mustafasariel** [comment]
I updated the SolrNet.Core library. 1.0.19->1.1.1, but the problem still persists.

**Comment by kanxingyv** [comment]
> Currently accepting PRs with a test+fix for this. I don't even know if it's SolrNet at fault here. As I said before I have tested unicode scenarios since 2009 with no issues. It's emojis in particular that cause the issue it seems.

@mausch  Could it be possible that SolrNet uses XML submission by default, while @HakanKaraoglu  second method uses JSON, which might handle emojis better? Have you tried configuring SolrNet to use application/json instead of text/xml for data submission? This might resolve the encoding issue with emojis.

**Comment by kanxingyv** [comment]
> I implemented a simple test that fails:
> 
> [Fact]
> public async Task AddAndQueryEmojis() {
>     var solr = ServiceLocator.Current.GetInstance<ISolrOperations<Product>>();
>     var id = Guid.NewGuid().ToString();
>     const string name = "Emojis are awesome! 😀😁😍😆😟😱😼😻🙌😀🤲😹🧠🤤";
>     System.Console.WriteLine(name);
>     await solr.AddAsync(new Product
>     {
>         Id = id,
>         Name = name,
>     });
>     await solr.CommitAsync();
>     try
>     {
>         var results = await solr.QueryAsync(new SolrQueryByField("id", id));
>         Assert.Single(results);
>         Assert.Equal(id, results[0].Id);
>         Assert.Equal(name, results[0].Name);
>     }
>     finally
>     {
>         await solr.DeleteAsync(id);
>     }
> }
> I tracked the issue down to this RegEx that removes the emojis:
> 
> [SolrNet/SolrNet/Impl/SolrDocumentSerializer.cs](https://github.com/SolrNet/SolrNet/blob/31672bad80038b2757259a74c8c9be4049c5cfcb/SolrNet/Impl/SolrDocumentSerializer.cs#L37)
> 
> Line 37 in [31672ba](/SolrNet/SolrNet/commit/31672bad80038b2757259a74c8c9be4049c5cfcb)
> 
>  new Regex(@"[^\x09\x0A\x0D\x20-\uD7FF\uE000-\uFFFD\u10000-u10FFFF]", RegexOptions.Compiled);

**Your analysis is spot-on!** 🎯 The regex’s exclusion of surrogate pairs (\uD800-\uDFFF) and the incorrect syntax for high Unicode ranges (like \u10000-u10FFFF) are exactly why both emojis and CJK Extension B characters (e.g., 𨿽 U+28FFD) are being stripped. I actually stumbled upon this issue precisely because my application needs to support rare CJK characters from Extension B (U+20000-U+2A6DF), which also rely on surrogate pairs — the same mechanism emojis use!
